lib/rules/media-feature-name-unit-allowed-list/index.js

```diff
@@ -0,0 +1,97 @@
+'use strict';
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+const { TokenType } = require('@csstools/css-tokenizer');
+const { isTokenNode } = require('@csstools/css-parser-algorithms');
+const {
+	isMediaFeaturePlain,
+	isMediaFeatureRange,
+	parse,
+} = require('@csstools/media-query-list-parser');
+const validateObjectWithArrayProps = require('../../utils/validateObjectWithArrayProps');
+const { isString } = require('../../utils/validateTypes');
+const matchesStringOrRegExp = require('../../utils/matchesStringOrRegExp');
+const flattenArray = require('../../utils/flattenArray');
+const atRuleParamIndex = require('../../utils/atRuleParamIndex');
+
+const ruleName = 'media-feature-name-unit-allowed-list';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (unit, name) => `Unexpected unit "${unit}" for name "${name}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/media-feature-name-unit-allowed-list',
+};
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [validateObjectWithArrayProps(isString)],
+		});
+
+		if (!validOptions) {
+			return;
+		}
+
+		root.walkAtRules(/^media$/i, (atRule) => {
+			const mediaQueryList = parse(atRule.params);
```

Comment:
[ask] If a parse error occurs from the `parse()` function, is it possible to recover the error like the `parseSelector()` utility?

https://github.com/stylelint/stylelint/blob/5e6d46aad5f61ef86f5823a67696a137a0e77a10/lib/utils/parseSelector.js#L13-L19

Reply:
This is not simple because the parser is forgiving and doesn't just throw whenever it encounters something unexpected.

Some details :
- the media query parser will never throw, if it does cause exceptions it would be a bug and must be fixed
- token stream level errors can be emitted
- component value level errors can be emitted
- there are no parser errors possible when applying the media query grammar to component values
- invalid media queries are component values that do not fit "todays" grammar
- invalid media queries can be preserved so that feedback can be given to authors

```ts
parse('something with parse errors', {
  onParseError: (err) => {
    // contains :
    // - tokenizer errors
    // - component value parser errors
    console.log(err);
  },
  preserveInvalidMediaQueries: true
});
```

Tokenizer/parser level errors are highly technical and will be rare.

Invalid media queries can not have a specific error messages because these are unknowns.
We only know that the component values do not fit the known grammar.

It is possible to emit a warning with `failed to parse X` but it is not possible to explain why it failed to parse it.

----

I suggest we handle this separately from this pull request?
Ideally we have good examples of error cases and the messages we want for each.

Then this work can be done once for all rules that would use this parser :)

Reply:
Thanks for the explanation. I understand well. As the first step, we can proceed as-is without considering edge cases.

Or, we may have the following utility for other rules using `@csstools/media-query-list-parser` (this PR doesn't need to add it, though)

```js
// lib/utils/parseMediaQuery.js

/**
 * @param {import('postcss').AtRule} atRule
 * @param {import('stylelint').PostcssResult} result
 * @returns {ReturnType<typeof parse>}
 */
module.exports = function parseMediaQuery(atRule, result) {
	const source = atRule.params;
	return parse(source, {
		onParseError: (err) => {
			result.warn(`Cannot parse selector (${err})`, {
				node: atRule,
				word: source,
				stylelintType: 'parseError',
			});
		},
		preserveInvalidMediaQueries: true,
	});
};
```

Reply:
If we'd like, I'm happy to open an issue to hammer out which cases need to be accounted for, and then draft a PR that uses this util stub + adds tests. Should I do that?

Reply:
Yes, please.

Reply:
I have created https://github.com/stylelint/stylelint/issues/6559; I'm not sure if I've captured the nuance of the request, feel free to chime in with more!